Proposal for stopping connectors

mimaison opened the pull request:
Proposal for https://github.com/strimzi/strimzi-kafka-operator/issues/8713

Kafka 3.5 introduced the STOPPED state for connectors via a new endpoint `PUT /connectors/<CONNECTOR>/stop`. Compared to PAUSED where the connector and tasks stay instantiated, in STOPPED only the configuration of the connector is kept but nothing is actually running.

A connector is PAUSED state still uses some resources but it's faster to resume. So using pause is great if you want to stop data flowing for a short duration. Otherwise it's interesting to fully stop the connector with the new API. Also the STOPPED state will be required to use the offset delete/reset endpoints that should come with Kafka 3.6.

scholzj commented on 052-connector-state.md:
```diff
@@ -0,0 +1,77 @@
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
+
+In the next Strimzi release the `pause` property will be marked as deprecated. In the following release, the `pause` field will be deleted.
```
You can mark it as deprecated together with this change. But it cannot be removed until the next API version.

ppatierno replied:
Should we also add a "plan" reporting Strimzi versions where these changes will happen?

mimaison replied:
Sorry I'm not familiar with the semantic versioning used by Strimzi. Is 0.40 the next API version change?

scholzj replied:
No. There is currently no exact Strimzi version when that happens. Next APi version will be basically when the `v1beta2` version of the custom resources moves to `v1` or `v1beta3`. I think it is unlikely to happen before KRaft is done and ZooKeeper gone.

scholzj commented on 052-connector-state.md:
````diff
@@ -0,0 +1,77 @@
+
+## Compatibility
+
+When first upgrading to a Strimzi version that adds `state`, `state` will be set to `pause` if `pause` is `true`, otherwise `state` will default to `run`. For example, upgrading with a connector like 
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  tasksMax: 1
+```
+will result in:
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  state: pause
+  tasksMax: 1
+```
````
We do not really do it this way as there is no simple way to modify the resource without risks to just end up in a loop with some other tools. So normally you would do something like this:
* The new field is preferred when both are set to some value
* When the old field is set in parallel with the new field, you will issue a warning about a conflcit
* When the new field is not set (i.e. is null), it will mean `run` if `pause == false` or if `pause` is not set. Or it will equal `pause` if `pause==true`. But this will not be done as a change to the custom resource. You will just handle it that way in the code.

This might stay like this until the next APi version which removes the `pause` field. Or you can say that both fields will be supported e.g. until Strimzi 0.40. In which case, in 0.41, we can change the code to completely ignore the `pause` field in the CRs ... but we still need to keep it in the custom resource definition as deprecated.

scholzj commented on 052-connector-state.md:
```diff
@@ -0,0 +1,77 @@
+
+## Motivation
+
+Since Kafka 3.5, Kafka Connect connectors can also be stopped. This is done by using a new REST endpoint `PUT /connectors/{connector}/stop`. Compared to paused where the connector and tasks stay instantiated, when stopped only the configuration of the connector is kept but nothing is actually running.
```
How will it be handled in Kafka 3.4?

mimaison replied:
Kafka 3.4 does not have this endpoint so connectors can't be stopped. Will the next Strimzi version still support Kafka < 3.5.0?

scholzj replied:
In general, we support two last minor Kafka versions. So once we add 3.6 support, we will likely drop 3.4 support. (I guess depending on Kafka schedules, but likely sometime shortly before the release of Strimzi 0.38)

I guess waiting is one option. Other option is simply documenting it as a known issue and handling it in the code so that it does not error.

mimaison replied:
Is it possible to prevent setting `state` to `stop` is the Kafka version is < 3.5? For example like what happens if you try setting `pause` to a value different than `true` or `false`?

scholzj replied:
Not really. The CRD exists only once. So you can prevent it from being set to `stop` -> but for all versions only.

mimaison replied:
I pushed an update to mention that if this is merged before support for Kafka 3.4 is dropped, the docs will mention it as a known issue.

scholzj reviewed (commented):
Thanks for the proposal. I think it is good in general -> I just left some comments abotu how API changes like this can be done.

ppatierno commented on 052-connector-state.md:
```diff
@@ -0,0 +1,77 @@
+
+## Proposal
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
```
Using a field in the spec let me think the different approach we took with the `KafkaRebalance` resource where we are using an annotation for pause, stop, refresh ...
Are we sure we want to go in a different direction here? 
Or we should change the `KafkaRebalance` to have the same approach?
It's more as a general discussion about setting a common pattern for the users from a UX pov about how to interact with custom resources which "run" operations which could be paused, stopped, restarted and so on.

scholzj replied:
I do not think this is the same. AFAIK most of the rebalance annotations are commands. This is a declarative management of the state.

katheris replied:
I agree with Jakub, the KafkaRebalance annotations are more commands to ask the operator to do something, rather than describing the current expected state of the resource.

In terms of the values, given the new field is called `state`, should the values be `paused`, `stopped`, and `running`

ppatierno replied:
Not sure, isn't it a command for the operator to invoke an API to Kafka Connect and ask to run an action (pause, stop and re-start a connector)?

scholzj replied:
No, it describes the desired state. Compare it for example with `strimzi.io/rebalance=refresh` => you are nto asking it to be refreshed every reconciliation (ie. constantly refreshing) but just once.

mimaison replied:
I'm happy to use `paused`, `stopped` and `running` as values. Then it's exactly the same as the `status.connectorStatus.connector.state` field which show the current state (apart as it's in uppercase in this field).

katheris replied:
@ppatierno @scholzj @tombentley what was your thought on changing the proposed values to `paused`, `stopped` and `running`, rather than `pause`, `stop` and `run`? I think the former makes more sense

scholzj replied:
I'm fine with `paused`, `stopped` and `running`. It should describe the desired state, so it probably fits better.

ppatierno replied:
+1 from me as well.

tombentley replied:
While I can't think of anywhere else that I've seen the gerund in resource-based APIs, I don't have a strong opinion. 

TBH the word `state` bothers me more. While it comes from the terminology used by connect itself, it's a bit more overloaded in a kube context, where we sometimes talk about the "desired state" (as a synonym for `spec`), and also use `status` as a way of communicating the connector's "state" more generally. I can't think of anything that is unambiguously better though. 
* Synonyms like `disposition` lose the connection to standard connect terminology
* Something like `connectorState` could be confused for a "KafkaConnector's `status`" . 
* Something like `targetState` might work (it's the term used in the KIP) but doesn't feel entirely natural because the `spec` is always expressing what's desired.

katheris replied:
I agree that the word `state` isn't ideal, however I do like that it aligns with the Connect output from the REST API and there isn't anything else that comes to mind as being a better fit.

katheris commented on 052-connector-state.md:
```diff
@@ -0,0 +1,53 @@
+- default to `pause` if `pause` is `true`
+But this will not be done as a change to the custom resource, it will just be handled that way internally.
+
+If `state` is set it will take precedence over the old `pause `field. If both fields are set, a warning will be emitted to notify the user of a potential conflict.
```
For the warning is it worth both logging it and adding it as a condition to the KafkaConnector resource?

mimaison replied:
I'm not sure I fully understand. Can you clarify what you mean by `adding it as a condition to the KafkaConnector resource`?

im-konge replied:
I think that @katheris means something like 
```
  conditions:
  - lastTransitionTime: "xxxxxxxxxxx"
    message: 'Both `state` and `pause` fields were set, there is a possibility of a conflict'
    reason: ...
    status: "True"
    type: Ready
 ```

+1 from me for the adding it as a condition to KC resource

katheris replied:
Yes, the K8s docs discuss them here: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions

tombentley commented on 052-connector-state.md:
```diff
@@ -0,0 +1,55 @@
+
+# Support stopping connectors
+
+This proposal aims at adding support for stopping Kafka Connect connectors.
```
To join the dots, can you link to the KIP where this was discussed in Apache Kafka?

mimaison replied:
Added a link to KIP-875.

tombentley commented on 052-connector-state.md:
```diff
@@ -0,0 +1,55 @@
+
+Since Kafka 3.5, Kafka Connect connectors can also be stopped. This is done by using a new REST endpoint `PUT /connectors/{connector}/stop`. Compared to paused where the connector and tasks stay instantiated, when stopped only the configuration of the connector is kept but nothing is actually running.
+
+A connector paused still uses some resources but it's faster to resume. So using pause is great if you want to halt data flowing for a short duration. For longer durations it's interesting to fully stop the connector with the new API. Also the stopped state will be required to use the offset delete/reset endpoints that should come with Kafka 3.6. For these reasons, Strimzi should support both pau ...
```
I think we need slight more justification than "interesting". (I'm picking on your choice of words; I don't doubt the usefulness of the feature).
```suggestion
A paused connector still uses some resources but it's faster to resume. So using pause is well-suited if you want to halt data flowing for a short duration. For longer durations it could be beneficial to free up memory and other resources by fully stopping the connector with the new API. The stopped state will be required to use the offset delete/reset endpoints that should come with Kafka 3.6. For these reasons, Strimzi should support both pausing and stopping connectors.
```

tombentley commented on 052-connector-state.md:
```diff
@@ -0,0 +1,55 @@
+
+The `pause` property will be marked as deprecated. In the next Strimzi API version, the `pause` field will be deleted.
+
+This feature requires at least Kafka 3.5 to work. In case it is merged when Strimzi still support older releases, a note in the documentation will be added to notify of this limitation.
```
You should specify how the operator would handle `stop` in this case. Should it fail the reconciliation, or ignore the state entirely, or treat `stop` as `pause`? Personally I think treating stopped as paused would be acceptable in this case since that's functionally equivalent.

mimaison replied:
Good point. I took your suggestion and added that when trying to stop a connector with 3.4, it will be paused.

tombentley reviewed (approved):
LGTM! Thanks @mimaison.

scholzj commented on 052-connector-state.md:
```diff
@@ -0,0 +1,55 @@
+
+## Proposal
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
+
+The `pause` property will be marked as deprecated. In the next Strimzi API version, the `pause` field will be deleted.
+
+This feature requires at least Kafka 3.5 to work. In case it is merged when Strimzi still support older releases, a note in the documentation will be added to notify of this limitation. When trying to stop a connector with Kafka 3.4, the connector will instead be paused.
+
+This proposal does not intend to address deleting/resetting offsets. As explained in the motivation section, stopping connectors has value on its own. Adding support for deleting/resetting connector offsets will be tackled in a separate proposal once Strimzi adopts a Kafka version that supports this feature (expected to release in 3.6).
```
Why did the example YAML disappear? I thought that was useful.

mimaison replied:
My bad. I added it back.

scholzj reviewed (approved):
Thanks for the proposal.

ppatierno commented on 052-connector-state.md:
```diff
@@ -0,0 +1,66 @@
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
+
+The `pause` property will be marked as deprecated. In the next Strimzi API version, the `pause` field will be deleted.
```
I am not sure we can delete the field, it would make "invalid" any `KafkaConnector` already in place. We should just ignore it. @scholzj wdyt? Isn't it similar to the removal of  `jaeger` type for tracing?

scholzj replied:
In the next API version, we can delete both `pause` as well as Jaeger tracing.

ppatierno reviewed (approved):
LGTM. I left just a nit.

PaulRMellor commented on 052-connector-state.md:
```diff
@@ -0,0 +1,66 @@
+
+The `pause` property will be marked as deprecated. In the next Strimzi API version, the `pause` field will be deleted.
+
+This feature requires at least Kafka 3.5 to work. In case it is merged when Strimzi still support older releases, a note in the documentation will be added to notify of this limitation. When trying to stop a connector with Kafka 3.4, the connector will instead be paused.
```
```suggestion
This feature requires at least Kafka 3.5 to work. In case it is merged when Strimzi still supports older releases, a note in the documentation will be added to notify of this limitation. When trying to stop a connector with Kafka 3.4, the connector will instead be paused.
```

PaulRMellor reviewed (approved):
Nice, clear proposal.

mimaison commented:
Thanks everyone for the comments and suggestions. I've renamed the values for the `state` field to `paused`, `stopped` and `running`. I've also incorporated @katheris suggestion about updating the message of the resource condition in case both the old and new fields are set.

katheris reviewed (approved):
LGTM

scholzj commented:
This has now 6 binding and one non-binding +1 votes. So I guess we can maybe leave it till tomorrow's EOB and if there are no other comments we can close it as approved?
